### DISCUSSION TITLE
关于kernel下的communicator模块设计的一些疑问

### DISCUSSION THREAD

**Discussion by LunarStore** [sogou/workflow#1631] [Q&A]
1. 为什么不将客户端连接的代码和服务端处理连接的代码分开，反而写在一个文件里面，表示这样会让源码阅读者抓不住关心的重点！
2. （以自低向上的看）直到我看了WFServerTask才明白communicator的在调用session的handle时传入的state参数是对WFServerTask的实现是有影响的，这样依赖是不是太重了，如果一个开发人员想自己添加一个ServerTask，那首先还得搞懂communicator的实现，这对开发者会不会不太友好？

**Comment by Barenboim** [comment]
Client和Server共用一套线程池和连接调度器才是最高效的方式，这样才能尽可能做到各个线程负载的均衡。

Communicator里，client的session是主动调用的，server session是被动产生的。用户的service类实现一个new_session的虚函数，在接收到新的请求是，会产生一个server session。server session正常需要handle两次，第一次是接收了完整的request，进入TOREPLY状态，第二次是回复结束，一般是SUCCESS状态。
